inverted_index.py

```diff
@@ -0,0 +1,354 @@
+    """
+    def ResetFiles(self):
+
+        block_dir = "blocks"
+        if os.path.exists(block_dir):
+            # Delete block folder
+            shutil.rmtree(block_dir)
+        os.makedirs(block_dir)
+
+        if os.path.exists(self.out_postings):
+            os.remove(self.out_postings)
+
+        if os.path.exists(self.out_dict):
+            os.remove(self.out_dict)
+
+    """
+            Method to write the contents of posting lists (stored as self.postings in memory) to a new block file
+            SPIMI-Invert
+            Params:
+                - block_index: Gives us the file name
+    """
+    def WriteBlockToDisk(self, block_index):
+        print("create new block ... " + str(block_index))
+
+        result = ""
+        # Sort Dictionary Terms. Line 11 of SPIMI - INVERT
+        for term in sorted(self.dictionary):
+            result += term + " "
+            # No need to sort postings because they are already in sorted form when we processed
+            # the docs in ascending order previously
+            for doc_id in self.postings[term]:
+                result += doc_id + " "
+            result += "\n"
+
+        self.WriteToFile("blocks/"+str(block_index), result)
+
+        # Clear Postings and Dictionary from memory
+        self.postings = defaultdict(list)
+        self.dictionary = set()
+
+    """
+            Method to read all the block files inside /Blocks and append them into Posting.txt
+            using BSBI Merging -> (total_num_blocks) K-Way Merge
+            Params:
+                - total_num_blocks: Total number of block files that we want to merge
+    """
+    def MergeBlocks(self, total_num_blocks):
+        print("Merge all blocks ...")
+
+        # blocks_offset[i] stores the offset value for reading from file of block i
+        blocks_offset = [0 for x in range(0, total_num_blocks+1)] # ignore value at index 0
+
+        q = PriorityQueue()
+
+        # Add the first line of each block to priority queue
+        for block_num in range(1, total_num_blocks + 1):
+            line = self.ReadFromFile("blocks/"+str(block_num), blocks_offset[block_num])
+            blocks_offset[block_num] = blocks_offset[block_num] + len(line) + 1 # 1 for \n
+            q.put(QueueItem(line, block_num))
+
+        # K-Way
+        term_to_write = ''
+        doc_ids_to_write = []
+
+        while not q.empty():
+
+            # Process current Item in the queue
+            curr_item = q.get()
+            curr_term = curr_item.GetTerm()
+            curr_posting_list = curr_item.GetLine().split(" ")[1:-1]
+            curr_block = curr_item.GetBlockNum()
+
+            # Encountering new term -> need to append previous term and it's doc_ids to posting.txt
+            if curr_term != term_to_write and term_to_write != '':
+                content = term_to_write + " " + " ".join(doc_ids_to_write)
+                self.WriteToFile(self.out_postings, content + "\n", True)
```

Comment:
same thing here as read, can keep the reference so that we don't have to keep opening the file and save on I/O cost. Also, we should write when buffer is full instead for every term?

Reply:
Correct...currently i haven't figured out how to set the buffer memory size so i implement it this way. Maybe i should add a TODO here.